Multihead support

chrisduerr opened the pull request:
Add method that makes it possible to get x-offset, y-offset, width,
height and some more properties of any specific output (by name).
This uses the root window, so it only needs connection and status to get
the info for the output. If the output cannot be found, it will return
an error.

This should make it very easy to fix #18. Which I can work on tomorrow. This already works and can be used. It just has to be connected to the rest. The user also has to be forced to specify an output that he wants to render the bar on.

dogamak commented:
Is there some reason you are using the C functions directly, and not the wrappers? You should use the wrappers everywhere you can and resort to `xcb::ffi::xcb_` functions only when the wrappers fail. The `xcb::ffi::xcb_randr_` functions should have more idiomatic and safer counterparts in `xcb::randr::`.

Also, did you notice you are using the `fix-17` branch? If you modify `calcuate_geometry` to use this function or do anything else related to multihead support, please add the commits to this PR and don't create a new one.

chrisduerr commented:
I haven't used the `xcb::randr::` bindings because I didn't know those exist. I simply followed one of the examples in the xcb github repository. I'll change that.

And using `fix-17` is intentional because I thought #17 would be fixed once Slot::Center and Slot::Right work. So I was gonna use this to fix those issues.

chrisduerr commented:
I've reworked the PR and removed all direct ffi calls except for `xcb::ffi::xproto::xcb_setup_roots_iterator(setup.ptr);` (I couldn't find an equivalent in the `xcb::xproto` docs.

dogamak commented:
The function `screen.roots()` is the equivalent of `xcb_setup_roots_iterator(screen.ptr)`. Preferably return `(u16, u16)` and use the `_checked` variants of the XCB funtions wrapped with `try_xcb!()`. Returning a tuple allows easier implementation of possible future Xinerama support.

*Tip: grep the source code of `rust-xcb` with C function names to find the equivalent rust functions.*

chrisduerr commented:
It looks to me like the XCB functions used are checked by default. For `get_output_info` for example there only is `get_output_info_unchecked` available, not `get_output_info_checked`.

I will add the `screen.roots()` call though.

chrisduerr commented:
I've added the `screen-roots()` fix. If my assumption that the other methods are checked by default is wrong, I can add that too. I'd also assume that `try_ccb!()` is only for those explicit checked methods. So I haven't meddled with that.

It now all works without unsafe. So at least I'm getting somewhere. :D

chrisduerr commented:
So is this ready for merge or is there still something I can improve? If you want I can also already add the configuration required to make this work. So that the user specifies the output when creating the bar. I'm already working on that and it would fully make multihead work. (I think)

Even tho I haven't thought about Relative and Absolute yet.

This already requires me to render on the specified screen tho. Which might include a little more work too. But I can look into it.

chrisduerr commented:
I've just added the changes required to make this work properly on multihead. With `Position::Top` this now works on all my outputs. However `Position::Bottom` doesn't work at all. I'm not entirely sure why and it probably never worked for me. The `y` calculated is certainly accurate.

If I should create separate PRs or something should be changed, please tell me.
I've decided to require the output to be specified inside of `BarBuilder::new` because it currently is required to do anything. However it would be possible to use the primary output as default. I'm not quite sure if it is worth implementing that tho because it adds a bit of complexity and doesn't really improve things that much imo.

chrisduerr commented:
@dogamak Do you have any feedback on the current state of this? I think it's as far as I would go in this PR.

dogamak commented:
Yeah, I think this is OK. Defining the output name would be preferably optional, but I'll just open an issue about it for now.

chrisduerr commented:
I'd also say since #17 is definitely fixed now it would probably be possible to merge fix-17 into master and remove that branch.
